### DISCUSSION TITLE
ResponseNotRead when using raise_for_status hook

### DISCUSSION THREAD

**Discussion by nicklas-delphai** [encode/httpx#1973] [Potential Issue]
In my setup I configure the `AsyncClient` to call `response.raise_for_status()` for every call using the `raise_on_4xx_5xx` response hook as [stated in the documentation](https://www.python-httpx.org/advanced/#event-hooks).

When catching `HTTPStatusError` as `ex` I was very confused to find that `ex.response.json()` would raise `ResponseNotRead` exception instead of giving me the expected json data.

After doing some digging in the source code I realized two things:
- The hooks are called right after the request was made [here](https://github.com/encode/httpx/blob/82ba15b52189eacec74f2ec7dbd81f3c70a69261/httpx/_client.py#L956-L957)
- This is before [response.read()](https://github.com/encode/httpx/blob/82ba15b52189eacec74f2ec7dbd81f3c70a69261/httpx/_client.py#L896-L897) is called for non-streaming calls

This will mean that if a user uses the `raise_on_4xx_5xx` as shown in the documentation they might unexpectedly run into the same issue as I did, where `read()` has not been called yet.

I believe this is very confusing. At least I can say it confused the hell out of me and cost me the better part of 2 hours now to figure out.

My solution to this problem was to extend the `raise_on_4xx_5xx` hook:

```py
async def raise_on_4xx_5xx(response: httpx.Response):
    if response.status_code >= 400:
        await response.aread()
    response.raise_for_status()
```

Also it's worth to note that I didn't need this workaround previous to v0.19. Before v0.19 everything worked as I would expect and I could call `ex.response.json()` without encountering this exception.

***Edit***

I probably should have read the note in the documentation:

> Response event hooks are called before determining if the response body should be read or not.
> If you need access to the response body inside an event hook, you'll need to call response.read(), or for AsyncClients, response.aread().

That's on me. Anyway, I still find it pretty unpredictable.

**Comment by lovelydinosaur** [comment]
Thanks Nicklas, that's useful feedback - might be worth us looking at if it's possible to improve on the user experience here one way or another.
